### PR TITLE
sessions: signout bug , fixes #530

### LINKED ISSUE #530: authenticate service crashing on logout after upgrade from 0.5.2 to 0.6.0
## What happened?

I bumped the version of the helm chart from 4.1.4 (pomerium 0.5.2) to 6.0.0 (pomerium 0.6.0) and while everything seems to work as expected, when I do a logout, I get the following error coming from the authenticate service in my browser (on the url `/.pomerium/sign_out?xxxxx`): 
```
upstream connect error or disconnect/reset before headers. reset reason: connection termination
```

And the following in the logs of the authenticate service:
```json
{
    "level": "error",
    "addr": ":443",
    "time": "2020-03-08T14:45:29Z",
    "message": "http: panic serving 10.242.2.3:48750: runtime error: invalid memory address or nil pointer dereference\ngoroutine 242 [running]:\nnet/http.(*conn).serve.func1(0xc0000cba40)\n\t/usr/local/go/src/net/http/server.go:1767 +0x139\npanic(0xdf66a0, 0x1724680)\n\t/usr/local/go/src/runtime/panic.go:679 +0x1b2\ngithub.com/pomerium/pomerium/internal/identity.(*GoogleProvider).Revoke(0xc0001d4140, 0x10a1920, 0xc000363ce0, 0x0, 0x0, 0xe95720)\n\t/go/src/github.com/pomerium/pomerium/internal/identity/google.go:103 +0x93\ngithub.com/pomerium/pomerium/authenticate.(*Authenticate).SignOut(0xc0000ab6c0, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000367700, 0x46802c, 0xdefa40)\n\t/go/src/github.com/pomerium/pomerium/authenticate/handlers.go:205 +0x174\ngithub.com/pomerium/pomerium/internal/httputil.HandlerFunc.ServeHTTP(0xc00021b250, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000367700)\n\t/go/src/github.com/pomerium/pomerium/internal/httputil/handlers.go:40 +0x48\ngithub.com/pomerium/pomerium/authenticate.(*Authenticate).VerifySession.func1(0x7f1aec8d5ac8, 0xc00037a540, 0xc000367700, 0x40c698, 0x80)\n\t/go/src/github.com/pomerium/pomerium/authenticate/handlers.go:90 +0x213\ngithub.com/pomerium/pomerium/internal/httputil.HandlerFunc.ServeHTTP(0xc0001d4240, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000367700)\n\t/go/src/github.com/pomerium/pomerium/internal/httputil/handlers.go:40 +0x48\ngithub.com/pomerium/pomerium/internal/sessions.retrieve.func1.1(0x7f1aec8d5ac8, 0xc00037a540, 0xc000367600)\n\t/go/src/github.com/pomerium/pomerium/internal/sessions/middleware.go:31 +0x1a1\nnet/http.HandlerFunc.ServeHTTP(0xc0003634d0, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000367600)\n\t/usr/local/go/src/net/http/server.go:2007 +0x44\ngithub.com/pomerium/pomerium/internal/middleware.ValidateSignature.func1.1(0x7f1aec8d5ac8, 0xc00037a540, 0xc000367500, 0x0, 0x0)\n\t/go/src/github.com/pomerium/pomerium/internal/middleware/middleware.go:37 +0x217\ngithub.com/pomerium/pomerium/internal/httputil.HandlerFunc.ServeHTTP(0xc000363500, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000367500)\n\t/go/src/github.com/pomerium/pomerium/internal/httputil/handlers.go:40 +0x48\ngithub.com/rs/cors.(*Cors).Handler.func1(0x7f1aec8d5ac8, 0xc00037a540, 0xc000367500)\n\t/go/pkg/mod/github.com/rs/cors@v1.7.0/cors.go:219 +0x1b8\nnet/http.HandlerFunc.ServeHTTP(0xc0001d42c0, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000367500)\n\t/usr/local/go/src/net/http/server.go:2007 +0x44\ngithub.com/pomerium/csrf.(*csrf).ServeHTTP(0xc00020e0e0, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000367400)\n\t/go/pkg/mod/github.com/pomerium/csrf@v1.6.2-0.20190918035251-f3318380bad3/csrf.go:286 +0x511\ngithub.com/pomerium/pomerium/internal/middleware.SetHeaders.func1.1(0x7f1aec8d5ac8, 0xc00037a540, 0xc000367200)\n\t/go/src/github.com/pomerium/pomerium/internal/middleware/middleware.go:22 +0x388\nnet/http.HandlerFunc.ServeHTTP(0xc0001d42e0, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000367200)\n\t/usr/local/go/src/net/http/server.go:2007 +0x44\ngithub.com/gorilla/mux.(*Router).ServeHTTP(0xc000478240, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000367000)\n\t/go/pkg/mod/github.com/gorilla/mux@v1.7.3/mux.go:212 +0xe2\ngithub.com/pomerium/pomerium/internal/middleware.Healthcheck.func1.1(0x7f1aec8d5ac8, 0xc00037a540, 0xc000366f00)\n\t/go/src/github.com/pomerium/pomerium/internal/middleware/middleware.go:72 +0x4b0\nnet/http.HandlerFunc.ServeHTTP(0xc00037a1c0, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000366f00)\n\t/usr/local/go/src/net/http/server.go:2007 +0x44\ngithub.com/pomerium/pomerium/internal/log.RequestIDHandler.func1.1(0x7f1aec8d5ac8, 0xc00037a540, 0xc000366f00)\n\t/go/src/github.com/pomerium/pomerium/internal/log/middleware.go:157 +0x133\nnet/http.HandlerFunc.ServeHTTP(0xc00037a200, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000366e00)\n\t/usr/local/go/src/net/http/server.go:2007 +0x44\ngithub.com/pomerium/pomerium/internal/log.RefererHandler.func1.1(0x7f1aec8d5ac8, 0xc00037a540, 0xc000366e00)\n\t/go/src/github.com/pomerium/pomerium/internal/log/middleware.go:113 +0x190\nnet/http.HandlerFunc.ServeHTTP(0xc000273350, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000366e00)\n\t/usr/local/go/src/net/http/server.go:2007 +0x44\ngithub.com/pomerium/pomerium/internal/log.UserAgentHandler.func1.1(0x7f1aec8d5ac8, 0xc00037a540, 0xc000366e00)\n\t/go/src/github.com/pomerium/pomerium/internal/log/middleware.go:97 +0x190\nnet/http.HandlerFunc.ServeHTTP(0xc0002733b0, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000366e00)\n\t/usr/local/go/src/net/http/server.go:2007 +0x44\ngithub.com/pomerium/pomerium/internal/log.RemoteAddrHandler.func1.1(0x7f1aec8d5ac8, 0xc00037a540, 0xc000366e00)\n\t/go/src/github.com/pomerium/pomerium/internal/log/middleware.go:81 +0xc4\nnet/http.HandlerFunc.ServeHTTP(0xc000273410, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000366e00)\n\t/usr/local/go/src/net/http/server.go:2007 +0x44\ngithub.com/pomerium/pomerium/internal/log.HeadersHandler.func1.1(0x7f1aec8d5ac8, 0xc00037a540, 0xc000366e00)\n\t/go/src/github.com/pomerium/pomerium/internal/log/middleware.go:190 +0x10f\nnet/http.HandlerFunc.ServeHTTP(0xc000273470, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000366e00)\n\t/usr/local/go/src/net/http/server.go:2007 +0x44\ngithub.com/pomerium/pomerium/internal/middleware.SetHeaders.func1.1(0x7f1aec8d5ac8, 0xc00037a540, 0xc000366d00)\n\t/go/src/github.com/pomerium/pomerium/internal/middleware/middleware.go:22 +0x388\nnet/http.HandlerFunc.ServeHTTP(0xc0001d40c0, 0x7f1aec8d5ac8, 0xc00037a540, 0xc000366d00)\n\t/usr/local/go/src/net/http/server.go:2007 +0x44\ngithub.com/pomerium/pomerium/internal/log.AccessHandler.func1.1(0x109f6e0, 0xc0003880a0, 0xc000366d00)\n\t/go/src/github.com/pomerium/pomerium/internal/log/middleware.go:168 +0x1dc\nnet/http.HandlerFunc.ServeHTTP(0xc0001d40e0, 0x109f6e0, 0xc0003880a0, 0xc000366d00)\n\t/usr/local/go/src/net/http/server.go:2007 +0x44\ngithub.com/pomerium/pomerium/internal/log.NewHandler.func1.1(0x109f6e0, 0xc0003880a0, 0xc000366c00)\n\t/go/src/github.com/pomerium/pomerium/internal/log/middleware.go:23 +0x3b7\nnet/http.HandlerFunc.ServeHTTP(0xc00024c1c0, 0x109f6e0, 0xc0003880a0, 0xc000366c00)\n\t/usr/local/go/src/net/http/server.go:2007 +0x44\ngo.opencensus.io/plugin/ochttp.(*Handler).ServeHTTP(0xc00057fa60, 0x109f6e0, 0xc0003880a0, 0xc000366a00)\n\t/go/pkg/mod/go.opencensus.io@v0.22.2/plugin/ochttp/server.go:92 +0x2b9\ngithub.com/pomerium/pomerium/internal/telemetry/metrics.HTTPMetricsHandler.func1.1(0x109e620, 0xc00020e000, 0xc000366900)\n\t/go/src/github.com/pomerium/pomerium/internal/telemetry/metrics/http.go:131 +0x35a\nnet/http.HandlerFunc.ServeHTTP(0xc000273590, 0x109e620, 0xc00020e000, 0xc000366900)\n\t/usr/local/go/src/net/http/server.go:2007 +0x44\ngithub.com/gorilla/mux.(*Router).ServeHTTP(0xc000478000, 0x109e620, 0xc00020e000, 0xc000366500)\n\t/go/pkg/mod/github.com/gorilla/mux@v1.7.3/mux.go:212 +0xe2\nnet/http.serverHandler.ServeHTTP(0xc00020e1c0, 0x109e620, 0xc00020e000, 0xc000366500)\n\t/usr/local/go/src/net/http/server.go:2802 +0xa4\nnet/http.(*conn).serve(0xc0000cba40, 0x10a1860, 0xc0003570c0)\n\t/usr/local/go/src/net/http/server.go:1890 +0x875\ncreated by net/http.(*Server).Serve\n\t/usr/local/go/src/net/http/server.go:2928 +0x384"
}
```

There are no problem if I start from a fresh session from within private browsing: I can login and logout without any problem, so I suppose it's coming from the cookie on my browser.

### PR DESCRIPTION
**Pull request by desimone**
To be cherrypicked to v0.6.X

Fixes #530 

**Checklist**:
- [x] add related issues
- [ ] updated docs
- [ ] updated unit tests
- [ ] updated UPGRADING.md
- [x] ready for review

### CODE REVIEW

File: internal/sessions/state.go [hunk 1]
```diff
@@ -182,3 +183,24 @@ func (s *State) accessTokenHash() string {
...
+	}{
+		Alias: (*Alias)(s),
+	}
+	if err := json.Unmarshal(b, &t); err != nil {
+		return err
+	}
+	if t.AccessToken == nil {
+		t.AccessToken = t.OldToken
+	}
+	*s = *(*State)(t.Alias)
```
Comment (travisgroth) [inline internal/sessions/state.go]: Seems fine though a bit opaque.  

Soft suggestion - Would it be possible to put the old field in `State`, and just worry about copying `OldToken` into `AccessToken` (if present) here?  I'm not sure if there's a downside to that approach but the code would be a lot clearer here.

Comment (desimone) [reply internal/sessions/state.go]: I explored a few different options.

### Using a getter.

> Soft suggestion - Would it be possible to put the old field in State, and just worry about copying OldToken into AccessToken (if present) here? I'm not sure if there's a downside to that approach but the code would be a lot clearer here.

This was my first thought, and approach. Would have resulted in... hundreds loc changes. Broke tests, and touched so many unrelated areas of the code base I am reticent to go this approach for a hotfix.

### Using a custom type for access token only.

This was my 2nd approach. The problem I found here was I still needed to make a bunch of unrelated code changes and higher level type casts. 

```go
type State struct {
	AccessToken   *oauthToken 
}

type oauthToken struct {
	*oauth2.Token
}
func (oa *oauthToken) UnmarshalJSON(b []byte) error {
	t := &struct {
		NewToken *oauth2.Token `json:"act,omitempty"`          // > v0.5.0
		OldToken *oauth2.Token `json:"access_token,omitempty"` // < v0.5.0
	}{}
	if err := json.Unmarshal(b, &t); err != nil {
		return err
	}
	oa = &oauthToken{t.NewToken}
	if oa == nil {
		oa = &oauthToken{t.OldToken}
	}
	return nil
}```

Comment (travisgroth) [reply internal/sessions/state.go]: Ouch.  Well given it's due to be removed, best to keep it contained.

Comment (desimone) [reply internal/sessions/state.go]: I think approach 1. is probably the best, in general, but for now.. I think this will work.

**Pull request merged**